### PR TITLE
Paginate comments

### PR DESCRIPTION
**Pull request by aarek**

### CODE REVIEW

File: spec/services/paginated_comments_spec.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+require 'rails_helper'
+
+describe PaginatedComments do
+
```
Comment (aarek) [inline spec/services/paginated_comments_spec.rb]: remove blank line

File: spec/services/paginated_comments_spec.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  let(:user) { create(:user) }
```
Comment (aarek) [inline spec/services/paginated_comments_spec.rb]: user is not needed

File: spec/services/paginated_comments_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    before { create_list(:comment, comment_count, debate: debate, status: :accepted) }
+
+    it 'returns list of comments with limit' do
+      expect(subject[:comments].count).to eq(limit)
+      expect(subject).to include(:next_position => sorted_ids.first - limit)
+    end
+  end
+
+  context 'when everything was shown' do
+    let(:start_at) { sorted_ids.last }
```
Comment (aarek) [inline spec/services/paginated_comments_spec.rb]: I think that in this scope there won't be any comment

File: spec/services/paginated_comments_spec.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    it 'returns empty array' do
+      expect(subject[:comments]).to be_empty
+    end
+  end
```
Comment (aarek) [inline spec/services/paginated_comments_spec.rb]: i would add some tests to check if `PaginatedComments` returns correct results based on `params` passed. For example `it 'returns 10 comments if limit not passed/set`, `it 'returns last comments if position not passed'` etc. Also i would change these `context`s to better reflect what we are testing. Maybe something like this:

```
PaginatedComments
  returns specified number of comments
  returns comments from given position
  returns next position
  when position not specified
    returns comments from newest available
  when limit not specified
    returns 10 comments
  ...additionally some checks when position/limit does not cast to int properly
```

File: app/services/paginated_comments.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
+class PaginatedComments
+  DEFAULT_LIMIT = 10
+
+  def initialize(comments_relation:, params:, direction: :backwards)
+    @params = params
+    @comments_stream = ARStream.build(direction, comments_relation, start_at: position)
+  end
+
+  def to_h
```
Comment (aarek) [inline app/services/paginated_comments.rb]: I would rename this to `call`

File: app/services/paginated_comments.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  def limit
+    limit = params[:limit].to_i
+    return DEFAULT_LIMIT if limit.zero?
+
+    limit
+  end
+
+  def position
+    position = params[:position].to_i
+    return nil if position.zero?
```
Comment (aarek) [inline app/services/paginated_comments.rb]: I think that zero could be valid value here so we probably need a different check to return `nil`

File: spec/services/paginated_comments_spec.rb [hunk 7]
```diff
@@ -10,21 +10,24 @@
...
+  subject { described_class.new(comments_relation: debate.retrieve_comments,
+                                direction: :backward,
+                                params: params).to_h }
 
   context 'When comments are created' do
     before { create_list(:comment, comment_count, debate: debate, status: :accepted) }
 
-    it 'returns list of comments' do
-      expect(subject.count).to eq(limit)
+    it 'returns list of comments with limit' do
```
Comment (aarek) [inline spec/services/paginated_comments_spec.rb]: I would split this to two separate tests, one checking limit and other checking next position. Also i would check here if position is numeric only

File: spec/controllers/api/comments_controller_spec.rb [hunk 8]
```diff
@@ -109,7 +111,8 @@
       it 'retrieve valid status and empty array of comments' do
         expect(subject).to have_http_status(:ok)
         expect(json_response).to eq("debate_closed" => false,
-                                    "comments" => [])
+                                    "comments" => [],
+                                    "next_position" => nil)
```
Comment (aarek) [inline spec/controllers/api/comments_controller_spec.rb]: i think we should split these checks into separate tests

File: spec/controllers/api/comments_controller_spec.rb [hunk 9]
```diff
@@ -96,7 +97,8 @@
       it 'retrieve valid status and empty array of comments' do
         expect(subject).to have_http_status(:ok)
         expect(json_response).to eq({ "debate_closed" => false,
-                                      "comments" => [] })
+                                      "comments" => [],
+                                      "next_position" => nil })
```
Comment (aarek) [inline spec/controllers/api/comments_controller_spec.rb]: i think we should split these checks into separate tests

File: spec/controllers/api/comments_controller_spec.rb [hunk 10]
```diff
@@ -115,19 +96,64 @@
       end
 
       it 'retrieve valid status and empty array of comments' do
```
Comment (cezarykaras) [inline spec/controllers/api/comments_controller_spec.rb]: We don't check a status here

File: app/services/paginated_comments.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
+class PaginatedComments
+  DEFAULT_LIMIT = 10
+
+  def initialize(comments_relation:, params:, direction: :backwards)
+    @params = params
```
Comment (aarek) [inline app/services/paginated_comments.rb]: now that i am thinking about this, we don't need whole params hash `here`, maybe we should assign `limit` and `position` in constructor and check instance variable in getter methods, something like this:

```ruby
def initialize
  @limit    = params[:limit].to_i
  @position = params[:position]
end

def limit
  return DEFAULT_LIMIT if @limit.zero?
  @limit
end

...
```

File: spec/controllers/api/comments_controller_spec.rb [hunk 12]
```diff
@@ -75,36 +67,25 @@
...
 
+    it 'returns debate status in json' do
+      expect(json_response).to include('debate_closed' => false)
+    end
+
+    it 'has 200 ok http status' do
+      expect(subject).to have_http_status(:ok)
+    end
+
     context 'when comments are created and have status active' do
```
Comment (cezarykaras) [inline spec/controllers/api/comments_controller_spec.rb]: It's a small thing, but comments have status accepted, not active

File: spec/controllers/api/comments_controller_spec.rb [hunk 13]
```diff
@@ -114,20 +95,65 @@
         create(:comment, user: user, debate: debate, status: :pending)
```
Comment (cezarykaras) [inline spec/controllers/api/comments_controller_spec.rb]: We can use here create_list instead of creating two comments separately

**Review by cezarykaras — APPROVED** [review]

**Pull request merged**